### PR TITLE
[Model] Add `dgl.nn.CuGraphSAGEConv` model

### PR DESCRIPTION
**Pull request by tingyu66**
## Description

This PR adds a GraphSAGE model Add `dgl.nn.CuGraphSAGEConv` that uses the accelerated sparse aggregation primitives in cugraph-ops. It requires `pylibcugraphops >= 23.02`.

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] I've leverage the [tools](https://docs.google.com/document/d/1iHyj7zlmygKSk5gBPsqIqL5ASPzJSPREaNT_QdsiYA4/edit) to beautify the python and c++ code.
- [x] The PR is complete and small, read the [Google eng practice (CL equals to PR)](https://google.github.io/eng-practices/review/developer/small-cls.html) to understand more about small PR. In DGL, we consider PRs with less than 200 lines of core code change are small (example, test and documentation could be exempted).
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the best of my knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [x] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
- [x] New nn.Module: `dgl.nn.CuGraphSAGEConv`
- [x] Test that validates its results against `SAGEConv`

### Notes
Fixes rapidsai/cugraph-ops#177.

### CODE REVIEW

**Comment by Rhett-Ying** [comment]
@dgl-bot

**Comment by tingyu66** [comment]
Can I get a review for this PR? @mufeili  
The example is basically identical to `examples/pytorch/graphsage/node_classification.py`. We can remove it if you think it's redundant.

**Comment by mufeili** [comment]
@dgl-bot

File: docs/source/api/python/nn-pytorch.rst [hunk 1]
```diff
@@ -21,6 +21,7 @@ Conv Layers
     ~dgl.nn.pytorch.conv.EGATConv
     ~dgl.nn.pytorch.conv.EdgeConv
     ~dgl.nn.pytorch.conv.SAGEConv
+    ~dgl.nn.pytorch.conv.CuGraphSAGEConv
```
Comment (mufeili) [inline docs/source/api/python/nn-pytorch.rst:24]: Maybe consider having a new subsection like

```
CuGraph Conv Layers
----------------------------------------
```

File: python/dgl/nn/pytorch/conv/cugraph_sageconv.py [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+
+    This module depends on :code:`pylibcugraphops` package, which can be
+    installed via :code:`conda install -c nvidia pylibcugraphops>=23.02`.
+
+    .. note::
+        This is an **experimental** feature.
+
+    Parameters
+    ----------
+    in_feats : int
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_sageconv.py:32]: `SageConv` supports source and destination nodes with different feature size. I assume this is not the case for this implementation.

File: python/dgl/nn/pytorch/conv/cugraph_sageconv.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+        Input feature size.
+    out_feats : int
+        Output feature size.
+    aggregator_type : str
+        Aggregator type to use (``mean``, ``sum``, ``min``, ``max``).
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_sageconv.py:37]: The options seem to be different from the ones for `GraphConv`, which are mean, gcn, pool, lstm.

File: python/dgl/nn/pytorch/conv/cugraph_sageconv.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        Dropout rate on features, default: ``0``.
+    bias : bool
+        If True, adds a learnable bias to the output. Default: ``True``.
+
+    Examples
+    --------
+    >>> import dgl
+    >>> import torch
+    >>> from dgl.nn import CuGraphSAGEConv
+    ...
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_sageconv.py:48]: No need to have `...` here

File: python/dgl/nn/pytorch/conv/cugraph_sageconv.py [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        in_feats,
+        out_feats,
+        aggregator_type="mean",
+        feat_drop=0.0,
+        bias=True,
+    ):
+        if has_pylibcugraphops is False:
+            raise ModuleNotFoundError(
+                f"{self.__class__.__name__} requires pylibcugraphops >= 23.02 "
+                f"to be installed."
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_sageconv.py:76]: Might be better to also give the instruction to install `pylibcugraphops >= 23.02` here

File: python/dgl/nn/pytorch/conv/cugraph_sageconv.py [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+        super().__init__()
+        self.in_feats = in_feats
+        self.out_feats = out_feats
+        self.aggr = aggregator_type
+        self.feat_drop = nn.Dropout(feat_drop)
+        self.linear = nn.Linear(2 * in_feats, out_feats, bias=bias)
+
+    def reset_parameters(self):
+        r"""Reinitialize learnable parameters."""
+        self.linear.reset_parameters()
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_sageconv.py:95]: Previously `SageConv` considers Xavier uniform while `nn.Linear.reset_parameters` considers Kaiming uniform. I'm not sure about the effects of this difference.

File: python/dgl/nn/pytorch/conv/cugraph_sageconv.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+
+    def forward(self, g, feat, max_in_degree=None):
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_sageconv.py:97]: another difference, lack of support for `edge_weight`

File: python/dgl/nn/pytorch/conv/cugraph_sageconv.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        r"""Forward computation.
+
+        Parameters
+        ----------
+        g : DGLGraph
+            The graph.
+        feat : torch.Tensor
+            Node features. Shape: :math:`(|V|, D_{in})`.
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_sageconv.py:105]: maybe better to be consistent with the notations used in `GraphConv`, e.g., `|V|->N`

File: python/dgl/nn/pytorch/conv/cugraph_sageconv.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+                )
+                offsets_fg[: offsets.numel()] = offsets
+                offsets_fg[offsets.numel() :] = offsets[-1]
+
+                _graph = make_fg_csr(offsets_fg, indices)
+        else:
+            _graph = make_fg_csr(offsets, indices)
+
+        feat = self.feat_drop(feat)
+        h = SAGEConvAgg(feat, _graph, self.aggr)[: g.num_dst_nodes()]
```
Comment (mufeili) [inline python/dgl/nn/pytorch/conv/cugraph_sageconv.py:146]: Does this internally concatenate the aggregated messages and the original node features?

File: examples/advanced/cugraph/graphsage.py [hunk 10]
```diff
@@ -0,0 +1,200 @@
+import argparse
```
Comment (mufeili) [inline examples/advanced/cugraph/graphsage.py:1]: Did you run this script? If so, what performance number did you obtain?

File: examples/advanced/cugraph/graphsage.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        self.layers.append(CuGraphSAGEConv(hid_size, hid_size, "mean"))
+        self.layers.append(CuGraphSAGEConv(hid_size, out_size, "mean"))
+        self.dropout = nn.Dropout(0.5)
+        self.hid_size = hid_size
+        self.out_size = out_size
+
+    def forward(self, blocks, x):
+        h = x
+        for l, (layer, block) in enumerate(zip(self.layers, blocks)):
+            h = layer(block, h, max_in_degree=10)
```
Comment (mufeili) [inline examples/advanced/cugraph/graphsage.py:33]: This seems a bit hacky.

File: examples/advanced/cugraph/graphsage.py [hunk 12]
```diff
@@ -0,0 +1,200 @@
...
+
+
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--mode",
+        default="mixed",
+        choices=["cpu", "mixed", "puregpu"],
+        help="Training mode. 'cpu' for CPU training, 'mixed' for CPU-GPU mixed training, "
+        "'puregpu' for pure-GPU training.",
```
Comment (mufeili) [inline examples/advanced/cugraph/graphsage.py:174]: fix indent

**Review by mufeili — COMMENTED** [review]
done a pass

File: python/dgl/nn/pytorch/conv/cugraph_sageconv.py [hunk 13]
```diff
@@ -0,0 +1,149 @@
...
+    in_feats : int
+        Input feature size.
+    out_feats : int
+        Output feature size.
+    aggregator_type : str
+        Aggregator type to use (``mean``, ``sum``, ``min``, ``max``).
+    feat_drop : float
+        Dropout rate on features, default: ``0``.
+    bias : bool
+        If True, adds a learnable bias to the output. Default: ``True``.
```
[the remark this answers is not in the document]
Comment (tingyu66) [reply python/dgl/nn/pytorch/conv/cugraph_sageconv.py:46]: Equations added.

File: python/dgl/nn/pytorch/conv/cugraph_sageconv.py (hunk 9 above)
Comment (tingyu66) [reply python/dgl/nn/pytorch/conv/cugraph_sageconv.py:150]: Yes, exactly.

File: python/dgl/nn/pytorch/conv/cugraph_sageconv.py (hunk 6 above)
Comment (tingyu66) [reply python/dgl/nn/pytorch/conv/cugraph_sageconv.py:99]: I think Kaiming is more suitable here as ReLU is often the choice for the nonlinearity in GNN; Xavier was designed for sigmoid function.

File: examples/advanced/cugraph/graphsage.py (hunk 10 above)
Comment (tingyu66) [reply examples/advanced/cugraph/graphsage.py:1]: Yes, in terms of pure training time (not including dataloading), `SAGEConv` takes 2.5s per epoch, while `CuGraphSAGEConv` takes 2.0s, despite the overhead of coo-to-csc conversion. Test accuracy is also the same.

File: examples/advanced/cugraph/graphsage.py (hunk 11 above)
Comment (tingyu66) [reply examples/advanced/cugraph/graphsage.py:33]: Can you explain what needs to be done here? Are you suggesting to unpack to loop like this?
```python
h = F.relu(self.conv1(g[0], x))
h = F.relu(self.conv2(g[1], h))
...
```

File: examples/advanced/cugraph/graphsage.py (hunk 12 above)
Comment (tingyu66) [reply examples/advanced/cugraph/graphsage.py:174]: This is automatically formatted by lintrunner. I removed the cpu mode, as it is not supported by the model

File: docs/source/api/python/nn-pytorch.rst [hunk 14]
```diff
@@ -43,6 +41,11 @@ Conv Layers
     ~dgl.nn.pytorch.conv.PNAConv
     ~dgl.nn.pytorch.conv.DGNConv
 
+CuGraph Conv Layers
+----------------------------------------
```
Comment (mufeili) [inline docs/source/api/python/nn-pytorch.rst:45]: Add an empty line below and add

```
.. autosummary::
    :toctree: ../../generated/
    :nosignatures:
    :template: classtemplate.rst
```

File: examples/advanced/cugraph/graphsage.py (hunk 11 above)
Comment (mufeili) [reply examples/advanced/cugraph/graphsage.py:33]: I meant the specification of `max_in_degree`.

File: docs/source/api/python/nn-pytorch.rst (hunk 14 above)
Comment (tingyu66) [reply docs/source/api/python/nn-pytorch.rst:45]: Ah, forgot to code over this snippet. Added now.

File: examples/advanced/cugraph/graphsage.py (hunk 11 above)
Comment (tingyu66) [reply examples/advanced/cugraph/graphsage.py]: I see and I do agree that it is not an ideal interface. We did not make `max_in_degree` an attribute of `CuGraphSAGEConv` since it is a property of the graph (i.e., `block`), rather than the model. I have removed it from the example as this flag is optional. 
In the meantime, we are improving our aggregation primitives to be more flexible to eventually ditch this option.

**Comment by tingyu66** [comment]
Thank you @mufeili  for the review. Here is a list of disparities between `CuGraphSAGEConv` and `SAGEConv`:
-  `SAGEConv` allows different feature dimensions for source and destination nodes
- They cover different aggregation types
- `CuGraphSAGEConv` does not support edge weights


Some preliminary performance numbers using the included example:
|mode| mixed (uva)| pure gpu |
|---------|------------------|--------------|
| CuGraphSAGEConv| 2.0 s | 1.2 s |
| SAGEConv | 2.5 s | 1.7 s |

(copied over from the review comment above for better visibility)

File: examples/advanced/cugraph/graphsage.py (hunk 12 above)
Comment (tingyu66) [reply examples/advanced/cugraph/graphsage.py:174]: Changes pushed.

**Review by mufeili — APPROVED** [review]

**Comment by mufeili** [comment]
@dgl-bot
